### ISSUE TITLE
Dropout is Slow

### ISSUE THREAD

**Issue by thomelane** [apache/incubator-mxnet#13825]
## Description

Adding dropout to network seems to reduce speed of forward pass substantially.

I've seen this a few times now, and the latest was occasion was on the discussion forum:
https://discuss.mxnet.io/t/training-speed-in-mxnet-is-nearly-2-5x-times-slower-than-pytorch. User's training was x2.5 slower than PyTorch originally, but reported a **6 times speedup by removing dropout**, ultimately making training faster than PyTorch. User also reported a significant reduction in memory usage.

Could be related to https://github.com/apache/incubator-mxnet/issues/12976.

## Environment info (Required)

User above is using MXNet version 1.3.1 on Ubuntu 16.04.5 with GTX1080.
And I have been able to replicate on AWS EC2 p3.2xlarge.
```
(mxnet_p36) ubuntu@ip-172-31-14-75:~$ python diagnose.py
----------Python Info----------
Version      : 3.6.5
Compiler     : GCC 7.2.0
Build        : ('default', 'Apr 29 2018 16:14:56')
Arch         : ('64bit', '')
------------Pip Info-----------
Version      : 10.0.1
Directory    : /home/ubuntu/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/pip
----------MXNet Info-----------
Version      : 1.3.1
Directory    : /home/ubuntu/anaconda3/envs/mxnet_p36/lib/python3.6/site-packages/mxnet
Commit Hash   : 19c501680183237d52a862e6ae1dc4ddc296305b
----------System Info----------
Platform     : Linux-4.4.0-1074-aws-x86_64-with-debian-stretch-sid
system       : Linux
node         : ip-172-31-14-75
release      : 4.4.0-1074-aws
version      : #84-Ubuntu SMP Thu Dec 6 08:57:58 UTC 2018
----------Hardware Info----------
machine      : x86_64
processor    : x86_64
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                8
On-line CPU(s) list:   0-7
Thread(s) per core:    2
Core(s) per socket:    4
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
Stepping:              1
CPU MHz:               2699.984
CPU max MHz:           3000.0000
CPU min MHz:           1200.0000
BogoMIPS:              4600.16
Hypervisor vendor:     Xen
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              46080K
NUMA node0 CPU(s):     0-7
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single kaiser fsgsbase bmi1 hle avx2smep bmi2 erms invpcid rtm rdseed adx xsaveopt
----------Network Test----------
Setting timeout: 10
Timing for MXNet: https://github.com/apache/incubator-mxnet, DNS: 0.0019 sec, LOAD: 0.3825 sec.
Timing for Gluon Tutorial(en): http://gluon.mxnet.io, DNS: 0.0998 sec, LOAD: 0.0959 sec.
Timing for Gluon Tutorial(cn): https://zh.gluon.ai, DNS: 0.5308 sec, LOAD: 0.2854 sec.
Timing for FashionMNIST: https://apache-mxnet.s3-accelerate.dualstack.amazonaws.com/gluon/dataset/fashion-mnist/train-labels-idx1-ubyte.gz, DNS: 0.0082 sec, LOAD: 0.1051 sec.
Timing for PYPI: https://pypi.python.org/pypi/pip, DNS: 0.0099 sec, LOAD: 0.3274 sec.
Timing for Conda: https://repo.continuum.io/pkgs/free/, DNS: 0.0109 sec, LOAD: 0.0471 sec.
```

Package used (Python/R/Scala/Julia): Python

### Using `mx.nd.Dropout`

```
%%timeit
data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
for i in range(100):
    # using mode='always' to force dropout behaviour
    data = mx.nd.Dropout(data, 0.5, mode='always')
mx.nd.waitall()
```
```
# 1.44 s ± 512 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Approx ~4.5 times slower than custom dropout.

### Using Custom Dropout

```
%%timeit
data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
for i in range(100):
    dropout_mask = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu()) > 0.5
    data = data * dropout_mask
mx.nd.waitall()
```
```
# 325 ms ± 338 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Approx ~4.5 times faster than `mx.nd.Dropout`.

### Network using `mx.gluon.nn.Dropout`

```
net = mx.gluon.nn.HybridSequential()
net.add(mx.gluon.nn.Conv2D(channels=32, kernel_size=3, strides=2))
net.add(mx.gluon.nn.BatchNorm())
net.add(mx.gluon.nn.Activation('relu'))
net.add(mx.gluon.nn.Dropout(0.5))
net.add(mx.gluon.nn.GlobalMaxPool2D())
net.add(mx.gluon.nn.Dense(units=1000))
net.initialize(ctx=mx.gpu())
```
```
%%timeit
for i in range(10):
    data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
    with mx.autograd.record():
        output = net(data)
    output.backward()
    mx.nd.waitall()
```
```
# 524 ms ± 882 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Approx 4 times slower than without.

### Network without `mx.gluon.nn.Dropout`

```
net = mx.gluon.nn.HybridSequential()
net.add(mx.gluon.nn.Conv2D(channels=32, kernel_size=3, strides=2))
net.add(mx.gluon.nn.BatchNorm())
net.add(mx.gluon.nn.Activation('relu'))
net.add(mx.gluon.nn.GlobalMaxPool2D())
net.add(mx.gluon.nn.Dense(units=1000))
net.initialize(ctx=mx.gpu())
```
```
%%timeit
for i in range(10):
    data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
    with mx.autograd.record():
        output = net(data)
    output.backward()
    mx.nd.waitall()
```
```
# 123 ms ± 187 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Approx 4 times faster than with.

**Comment by eric-haibin-lin** [comment]
Probably resolved by #13896

**Comment by roywei** [comment]
after #13896, seems only `mx.gluon.nn.Dropout(0.5)` has performance improvements, but not  mx.nd.Dropout(data, 0.5, mode='always')

following the above code sample:
### Using `mx.nd.Dropout`
1.43 s ± 293 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
### Using Custom Dropout
331 ms ± 411 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)

### Network using `mx.gluon.nn.Dropout`
128 ms ± 39.4 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

### Network without `mx.gluon.nn.Dropout`
128 ms ± 30.2 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

tested with both `pip install mxnet-cu92 --pre` and build from source on p3.2x

**Comment by szha** [comment]
@roywei by default cudnn_off is turned on. You need to turn it off to benefit from cudnn dropout.

**Comment by sandeep-krishnamurthy** [comment]
@thomelane - Can you please confirm if the issue is resolved with CUDNN support?
Or, this may be related to this - https://github.com/apache/incubator-mxnet/issues/14198

**Comment by yuxihu** [comment]
Tested with mxnet-cu90mkl==1.5.0b20190501 on g3.8xlarge instance.

Code example (same as above):
```
import mxnet as mx
import time

tic = time.time()

data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
for i in range(100):
    # using mode='always' to force dropout behaviour
    data = mx.nd.Dropout(data, 0.5, mode='always')
mx.nd.waitall()

print("mx.nd.Dropout: ", time.time() - tic)

tic = time.time()
data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
for i in range(100):
    dropout_mask = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu()) > 0.5
    data = data * dropout_mask
mx.nd.waitall()

print("Custom Dropout: ", time.time() - tic)

net = mx.gluon.nn.HybridSequential()
net.add(mx.gluon.nn.Conv2D(channels=32, kernel_size=3, strides=2))
net.add(mx.gluon.nn.BatchNorm())
net.add(mx.gluon.nn.Activation('relu'))
net.add(mx.gluon.nn.Dropout(0.5))
net.add(mx.gluon.nn.GlobalMaxPool2D())
net.add(mx.gluon.nn.Dense(units=1000))
net.initialize(ctx=mx.gpu())

tic = time.time()
for i in range(10):
    data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
    with mx.autograd.record():
        output = net(data)
    output.backward()
    mx.nd.waitall()

print("mx.gluon.nn.Dropout: ", time.time() - tic)

net = mx.gluon.nn.HybridSequential()
net.add(mx.gluon.nn.Conv2D(channels=32, kernel_size=3, strides=2))
net.add(mx.gluon.nn.BatchNorm())
net.add(mx.gluon.nn.Activation('relu'))
net.add(mx.gluon.nn.GlobalMaxPool2D())
net.add(mx.gluon.nn.Dense(units=1000))
net.initialize(ctx=mx.gpu())

tic = time.time()
for i in range(10):
    data = mx.nd.random.uniform(shape=(100, 3, 224, 224), ctx=mx.gpu())
    with mx.autograd.record():
        output = net(data)
    output.backward()
    mx.nd.waitall()

print("Without mx.gluon.nn.Dropout: ", time.time() - tic)
```

Performance results (in seconds):
```
mx.nd.Dropout:  2.713014841079712
Custom Dropout:  3.227015733718872
mx.gluon.nn.Dropout:  0.9665834903717041
Without mx.gluon.nn.Dropout:  0.7113254070281982
```
Looks like the dropout performance has been significantly improved with CUDNN support.

@sandeep-krishnamurthy Please help close this issue. 

@thomelane Please reopen the issue if you have any further questions. Thanks.
